src/opencl_argon2_fmt_plug.c

```diff
@@ -0,0 +1,791 @@
+	{"$argon2i$v=19$m=16384,t=4,p=3$YW5vdGhlcl9zYWx0$K7unxwO5aeuZCpnIJ06FMCRKod3eRg8oIRzQrK3E6mGbyqlTvvl47jeDWq/5drF1COJkEF9Ty7FWXJZHa+vqlf2YZGp/4qSlAvKmdtJ/6JZU32iQItzMRwcfujHE+PBjbL5uz4966A", "death_dying"},
+	{"$argon2id$v=19$m=4096,t=3,p=1$c2hvcmF0X3NhbHQ$K6/V3qNPJwVmLb/ELiD8gKGskLaFv5OweJYwSKUW1hE", "password"},
+	{"$argon2id$v=19$m=16384,t=4,p=3$c2hvcmF0X3NhbHQ$hG83oaWEcftTjbiWJxoQs6gKCModwYAC+9EK8j/DUsk", "sacrificed"},
+	{NULL}
+};
+// TODO: Backport many of the improvements to the CPU format
+// TODO: Optimize OpenCL code
+
+struct argon2_salt {
+	uint32_t t_cost, m_cost, lanes;
+	uint32_t hash_size;
+	uint32_t salt_length;
+	char salt[SALT_SIZE];
+	argon2_type type;
+	argon2_version version;
+};
+static struct argon2_salt saved_salt;
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1] = NULL;
+static int *saved_len = NULL;
+static uint8_t (*crypted)[BINARY_SIZE] = NULL;
+
+// GPU functions and memory
+#define ARGON2_NUM_TYPES 3
+static cl_int cl_error = CL_SUCCESS;
+static cl_kernel kernels[ARGON2_NUM_TYPES] = {NULL, NULL, NULL};
+static cl_mem memory_buffer = NULL;
+
+// CPU buffers to move data from and to the GPU
+static uint8_t* blocks_in = NULL;
+static uint8_t* blocks_out = NULL;
+
+// Autotune params
+struct kernel_run_params
+{
+	uint32_t lanes_per_block;
+	uint32_t jobs_per_block;
+};
+static struct kernel_run_params* best_kernel_params = NULL;
+static uint32_t max_salt_lanes = 0;
+static uint32_t max_segment_blocks = 0;
+
+#define THREADS_PER_LANE 32
+#define ARGON2_REFS_PER_BLOCK 	(ARGON2_BLOCK_SIZE / (2 * sizeof(cl_uint)))
+
+static int run_kernel_on_gpu(uint32_t lanes_per_block, size_t jobs_per_block)
+{
+        uint32_t pass, slice;
+        uint32_t lanes = saved_salt.lanes;
+        uint32_t passes = saved_salt.t_cost;
+
+	assert(lanes > 0 && passes > 0 && saved_salt.m_cost > 0);
+
+	if (lanes_per_block > lanes || lanes % lanes_per_block != 0)
+		error_msg("Invalid lanes_per_block!");
+
+	if (jobs_per_block > MAX_KEYS_PER_CRYPT || MAX_KEYS_PER_CRYPT % jobs_per_block != 0)
+    	error_msg("Invalid jobs_per_block!");
+
+	size_t global_range[2] = {THREADS_PER_LANE * lanes, MAX_KEYS_PER_CRYPT};
+	size_t local_range[2] = {THREADS_PER_LANE * lanes_per_block, jobs_per_block};
+
+	// Calculate memory size
+	uint32_t segment_blocks = MAX(saved_salt.m_cost / (saved_salt.lanes * ARGON2_SYNC_POINTS), 2);
+        // Copy data to GPU
+	size_t jobSize = segment_blocks * ARGON2_SYNC_POINTS * saved_salt.lanes * ARGON2_BLOCK_SIZE;
+	size_t copySize = saved_salt.lanes * 2 * ARGON2_BLOCK_SIZE;
+	size_t zero3[3] = {0, 0, 0};
+	size_t region3_in[3] = {copySize, MAX_KEYS_PER_CRYPT, 1};
+	assert(gpu_id >= 0 && gpu_id < MAX_GPU_DEVICES && queue[gpu_id]);
+	assert(blocks_in && blocks_out);
+	HANDLE_CLERROR(clEnqueueWriteBufferRect(queue[gpu_id], memory_buffer, CL_FALSE, 
+						zero3, zero3,
+						region3_in,
+						jobSize, 0, copySize, 0, blocks_in, 0, NULL, NULL), "Copy data to gpu");
```

Comment:
This is the only usage of `clEnqueueWriteBufferRect` currently in our tree. Is there any special reason why we need this specific OpenCL function here?

Reply:
Because we need to copy the memory to the GPU by pieces and not continuously, and that's what `clEnqueueWriteBufferRect` provides.